### PR TITLE
Create listener to refresh search thread resource usage

### PR DESCRIPTION
**Pull request by deshsidd**
Create listener (SearchTaskResourceOperationsListener) to refresh search thread resource usage. This will be primarily used by the query-insights plugin (https://github.com/opensearch-project/query-insights) to access task resource usage.

### Description
SearchTaskResourceOperationsListener subscriber for operations on search tasks resource usages.
 Listener ensures to refreshResourceStats on request end capturing the search task resource usage
 upon request completion. This listener API is public and ensures plugins can get accurate search tasks resource usage.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

Related discussion : https://github.com/opensearch-project/OpenSearch/pull/14681/files#r1678438231

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java
Comment (reta) [inline server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java:72]: Please revert this change (the problem will be solved differently)

File: server/src/main/java/org/opensearch/tasks/SearchTaskResourceOperationsListener.java
Comment (reta) [inline server/src/main/java/org/opensearch/tasks/SearchTaskResourceOperationsListener.java:29]: You don't need these stub methods, keep only `onRequestEnd`

File: server/src/main/java/org/opensearch/tasks/SearchTaskResourceOperationsListener.java
Comment (reta) [inline server/src/main/java/org/opensearch/tasks/SearchTaskResourceOperationsListener.java:22]:
```suggestion
public final class SearchTaskRequestOperationsListener extends SearchRequestOperationsListener {
```

Comment (reta) [reply server/src/main/java/org/opensearch/tasks/SearchTaskResourceOperationsListener.java:22]: (suggestion, I think it reflects that this is the request listener, not task listener)

File: server/src/main/java/org/opensearch/tasks/SearchTaskResourceOperationsListener.java
Comment (reta) [inline server/src/main/java/org/opensearch/tasks/SearchTaskResourceOperationsListener.java:20]: Please remove `@PublicApi`

File: server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java
Comment (reta) [inline server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java:74]: Accidental?

File: server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java
Comment (deshsidd) [reply server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java]: Any suggestions on another way we can solve this issue? @reta

Comment (reta) [reply server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java]: There is no issue to solve so far: as we have discussed, using `@InternalApi` is no-go, promoting the API to public will bring the questions:
a) how it is going to be used by plugins (and extensions)
b) what guarantees it will have from the execution perspective

Comment (deshsidd) [reply server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java]: The above change was required because of the following discussion : https://github.com/opensearch-project/OpenSearch/pull/14681/files#discussion_r1680221328

**Comment by reta** [comment]
@jed326 since you have expressed an interest (https://github.com/opensearch-project/OpenSearch/pull/14681), wanna take a look as well? Thank you.

**Review by reta — APPROVED** [review]

File: server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java
Comment (reta) [reply server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java]: I remember the discussion, but please refer to the previous comment (https://github.com/opensearch-project/OpenSearch/pull/14832#discussion_r1683438694)

**Comment by deshsidd** [comment]
Thanks for the review @reta. Will wait for the build to finish before you can merge

File: server/src/main/java/org/opensearch/action/search/TransportSearchAction.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/TransportSearchAction.java:438]: Do you need to create new `SearchTaskRequestOperationsListener` for every request? Should it be singleton instance?

Comment (jed326) [reply server/src/main/java/org/opensearch/action/search/TransportSearchAction.java:438]: +1, IIRC there's only 1 instance of the `TaskResourceTrackingService` per node and it looks like this is doing something very similar.

Comment (deshsidd) [reply server/src/main/java/org/opensearch/action/search/TransportSearchAction.java]: Made `SearchTaskRequestOperationsListener` as singleton in the latest

File: server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java
Comment (andrross) [inline server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java:31]: This static instance is weird. It makes the assumption that all calls to this method will pass in the same taskResourceTrackingService instance and will silently do the wrong thing if that assumption ever does not hold in the future. Why is this necessary? Are you sure this is better than the lock-free alternative of creating a new instance on every invocation?

**Review by jed326 — APPROVED** [review]

File: server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java
Comment (jainankitk) [reply server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java:31]: @andrross - I had similar thoughts. Please let me know what you think about - https://github.com/opensearch-project/OpenSearch/pull/14832#discussion_r1683547868

File: server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java
Comment (jainankitk) [reply server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java:22]: Having the variable final ensures, compiler itself throws error if try initialize it elsewhere

File: server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java
Comment (deshsidd) [reply server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java:31]: @andrross - Please let us know your thoughts

File: server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java
Comment (andrross) [reply server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java:34]: Can we make the SearchTaskRequestOperationsListener instance a member of TransportSearchAction? We should avoid this static state if at all possible.

Comment (andrross) [reply server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java:34]: And maybe put SearchTaskRequestOperationsListener in the same Java package as TransportSearchAction so it can be package private?

**Comment by andrross** [comment]
> This listener API is public

@deshsidd Is this still true? If not, please update the PR description and changelog entry as appropriate.

**Comment by deshsidd** [comment]
@andrross @jainankitk Please take a look. Thanks

**Review by jainankitk — APPROVED** [review]
LGTM!

File: server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java
Comment (jainankitk) [reply server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java]: IMO having SearchTaskRequestOperationsListener separately is okay, but making the instance member of TransportSearchAction to get rid of all the static stuff makes lot of sense Looks so much better now!

**Comment by ansjcy** [comment]
@deshsidd Thanks a lot for following up on the change! However we might be missing one change in this PR: https://github.com/opensearch-project/OpenSearch/pull/14782/files#diff-673f8e993795b1d79af3e0eaee5b2f7254817ce1e18970c74d2c33d9327d281bR72

File: server/src/main/java/org/opensearch/action/search/SearchTaskRequestOperationsListener.java
Comment (andrross) [inline server/src/main/java/org/opensearch/action/search/SearchTaskRequestOperationsListener.java:28]: Are there any significant performance implications to doing this refresh on every search? Has this been measured?

**Comment by andrross** [comment]
{"run-benchmark-test": "id_4"}

File: server/src/main/java/org/opensearch/action/search/SearchTaskRequestOperationsListener.java
Comment (jainankitk) [reply server/src/main/java/org/opensearch/action/search/SearchTaskRequestOperationsListener.java:28]: All the work seems in memory and once when request ends. Hence, I doubt it can have significant implication. @ansjcy / @deshsidd - Please correct me if I am wrong

**Comment by jainankitk** [comment]
> @deshsidd Thanks a lot for following up on the change! However we might be missing one change in this PR: https://github.com/opensearch-project/OpenSearch/pull/14782/files#diff-673f8e993795b1d79af3e0eaee5b2f7254817ce1e18970c74d2c33d9327d281bR72

@ansjcy - I looked at the change and it seems we are modifying the order of listeners list. IMO search operations listener logic should be independent of the execution order.

**Comment by andrross** [comment]
{"run-benchmark-test": "id_4"}

File: server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java
Comment (deshsidd) [reply server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java]: > @deshsidd Thanks a lot for following up on the change! However we might be missing one change in this PR: https://github.com/opensearch-project/OpenSearch/pull/14782/files#diff-673f8e993795b1d79af3e0eaee5b2f7254817ce1e18970c74d2c33d9327d281bR72

@ansjcy This change was reverted based on the following comment and thread (reopening the conversation). Let me know your thoughts and what other alternatives we have.

**Comment by ansjcy** [comment]
@jainankitk The goal is to refresh the thread resource usage before plugin listeners are executed so that the usage info would be up to date. We are adding a per request listener to do this, which means it has to be executed at the begining - because the sequence we add listeners into `SearchRequestOperationsCompositeListenerFactory.java` is the actual sequence the listeners are executed. 

I actually also have some concerns/doubts around this because like you mentioned `SearchRequestOperationsListener`s should be independent of each other. If we don't want to break this we might need to think about another place to inject the refresh logic. Let me know what do you think.

**Comment by jainankitk** [comment]
> I actually also have some concerns/doubts around this because like you mentioned SearchRequestOperationsListeners should be independent of each other. If we don't want to break this we might need to think about another place to inject the refresh logic. Let me know what do you think.

Yes, the listeners by definition must be independent of each other. What other approach do we have for addressing this?

**Comment by deshsidd** [comment]
> @jainankitk The goal is to refresh the thread resource usage before plugin listeners are executed so that the usage info would be up to date. We are adding a per request listener to do this, which means it has to be executed at the begining - because the sequence we add listeners into `SearchRequestOperationsCompositeListenerFactory.java` is the actual sequence the listeners are executed.
> 
> I actually also have some concerns/doubts around this because like you mentioned `SearchRequestOperationsListener`s should be independent of each other. If we don't want to break this we might need to think about another place to inject the refresh logic. Let me know what do you think.

@reta Any thoughts on the above since you initially commented to revert the change?

**Comment by rishabh6788** [comment]
There has been a few modifications on the benchmark workflow. Triggering again. 
Also, the[ baseline comparison dashboard](https://s12d.com/basline-dashboards) is ready for you to compare pull request results against latest commits merged onto mainline. 
@andrross

**Comment by rishabh6788** [comment]
{"run-benchmark-test": "id_4"}

**Comment by reta** [comment]
> @reta Any thoughts on the above since you initially commented to revert the change?

@deshsidd I am all in to look for a better place to have refresh logic and not involve any listeners.

**Comment by jainankitk** [comment]
> @jainankitk The goal is to refresh the thread resource usage before plugin listeners are executed so that the usage info would be up to date. We are adding a per request listener to do this, which means it has to be executed at the begining - because the sequence we add listeners into SearchRequestOperationsCompositeListenerFactory.java is the actual sequence the listeners are executed.

@ansjcy - Please correct me if I am wrong, but it seems you're concerned about the correctness of usage info. Ideally the search operation listeners should be lightweight and should not be adding much overhead (compared to the original search request execution). Do you believe that not to be the case?

**Comment by rishabh6788** [comment]
The benchmark results on the PR changes align with baseline numbers, https://s12d.com/basline-dashboards.

**Comment by jainankitk** [comment]
@rishabh6788 - Is there a good way to read the benchmark numbers - https://github.com/opensearch-project/OpenSearch/pull/14832#issuecomment-2243910637? Will be good to have a baseline column to compare against

**Comment by rishabh6788** [comment]
> @rishabh6788 - Is there a good way to read the benchmark numbers - [#14832 (comment)](https://github.com/opensearch-project/OpenSearch/pull/14832#issuecomment-2243910637)? Will be good to have a baseline column to compare against

This feature is in progress, https://github.com/opensearch-project/OpenSearch/issues/14826. Once completed, a final comparison report would also be published to the PR. 
Till then you can use the mentioned dashboard and use filters to focus on particular queries. 
I will also add a data table to help make it more readable.

File: server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java
Comment (deshsidd) [reply server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java]: Making a change to Node.java to make sure that the core listener runs before the plugin listener for now. Logically, it makes sense to make sure the core listeners are run before the plugin listeners. However, we will do this without making the above change and add the listener directly from Node.java. Discussed with @ansjcy @jainankitk @kkhatua. If needed we can make any additional fixes and refactoring as bug fixes.

**Review by jainankitk — APPROVED** [review]
LGTM!

File: server/src/main/java/org/opensearch/node/Node.java
Comment (ansjcy) [inline server/src/main/java/org/opensearch/node/Node.java:859]: We don't need to do this since `taskResourceTrackingService` is injected

**Comment by deshsidd** [comment]
Making a change to Node.java to make sure that the core listener runs before the plugin listener for now. Logically, it makes sense to make sure the core listeners are run before the plugin listeners. However, we will do this without switching the order of the current listeners and add the listener directly from Node.java (similar to what we did for slow logs). Discussed with @ansjcy @jainankitk @kkhatua. If needed we can make any additional fixes and refactoring as bug fixes. Hope we can merge this for now based on the above. 

cc @reta @andrross @mch2 @jed326

**Review by jed326 — APPROVED** [review]

**Pull request merged**
